File: drs/src/test/com/db4o/drs/test/db4o/PartialCollectionReplicationTestCase.java
/* Copyright (C) 2004 - 2010  Versant Inc.  http://www.db4o.com */

package com.db4o.drs.test.db4o;

import java.util.*;

import com.db4o.drs.db4o.*;
import com.db4o.drs.test.*;
import com.db4o.events.*;
import com.db4o.ext.*;
import com.db4o.foundation.*;

import db4ounit.*;

public class PartialCollectionReplicationTestCase extends DrsTestCase {
	
	public static class Data {
		private List _children;
		private String _id;
		
		public Data(String id) {
			_id = id;
			_children = new ArrayList();
		}
		
		public Object id() {
			return _id;
		}
		
		public void id(String id) {
			_id = id;
		}

		public void add(Data data) {
			_children.add(data);
		}
		
		@Override
		public String toString() {
			return "Data(" + _id + ", " + _children + ")";
		}
	}
	
	public void testNoReplicationForUntouchedElements() {
		final Data root = new Data("root");
		final Data c1 = new Data("c1");
		final Data c2 = new Data("c2");
		root.add(c1);
		root.add(c2);
		
		store(root, 1);
		
		final List<Data> created = replicateAllCapturingCreatedObjects();
		assertData(created, "root", "c1", "c2");
		
		final Data c3 = new Data("c3");
		root.add(c3);
		store(root, 2);
		
		c2.id("c2*");
		c2.add(new Data("c4"));
		
		final List<Data> updated = replicateAllCapturingUpdatedObjects();
		
		// The following fails after cleaning references has been removed from #replicate(obj) 
		// assertData(updated, "c3", "root");
	}

	private void assertData(final Iterable<Data> data, final String... expectedIds) {
		Iterator4Assert.sameContent(expectedIds, ids(data));
	}
	
	private Iterator4 ids(Iterable<Data> data) {
		final Collection4 ids = new Collection4();
		for (Data d : data) {
			ids.add(d.id());
		}
		return ids.iterator();
	}

	private List<Data> replicateAllCapturingUpdatedObjects() {
		final List<Data> changed = new ArrayList<Data>();
		listenToUpdated(changed);
		listenToCreated(changed);
		replicateAll();
		return changed;
	}
	
	private List<Data> replicateAllCapturingCreatedObjects() {
		final List<Data> created = new ArrayList<Data>();
		listenToCreated(created);
		replicateAll();
		return created;
	}

	private void listenToUpdated(final List<Data> updated) {
		eventRegistryFor(b()).updated().addListener(new EventListener4() {
			public void onEvent(Event4 e, EventArgs args) {
				final Object o = ((ObjectEventArgs)args).object();
				if (o instanceof Data) {
					updated.add((Data) o);
				}
				ods(o);
			}
		});
	}

	private void replicateAll() {
		ods("BEGIN REPLICATION");
		replicateAll(a().provider(), b().provider());
		ods("END REPLICATION");
	}

	private void listenToCreated(final List<Data> created) {
		eventRegistryFor(b()).created().addListener(new EventListener4() {
			public void onEvent(Event4 e, EventArgs args) {
				final Object o = ((ObjectEventArgs)args).object();
				if (o instanceof Data) {
					created.add((Data) o);
				}
				ods(o);
			}
		});
	}

	private EventRegistry eventRegistryFor(final DrsProviderFixture fixture) {
		return EventRegistryFactory.forObjectContainer(containerFor(fixture));
	}
	
	public void testCollectionUpdateDoesNotTouchExistingElements() {
		final Data root = new Data("root");
		final Data c1 = new Data("c1");
		final Data c2 = new Data("c2");
		root.add(c1);
		root.add(c2);
		
		store(root, 1);
		
		final long c1Version = versionFor(c1);
		final long c2Version = versionFor(c2);
		
		final Data c3 = new Data("c3");
		root.add(c3);
		store(root, 2);

		Assert.isGreater(0, versionFor(c3));
		Assert.areEqual(c1Version, versionFor(c1));
		Assert.areEqual(c2Version, versionFor(c2));
	}

	private void store(final Data root, final int depth) {
		final ExtObjectContainer container = containerFor(a());
		container.ext().store(root, depth);
		container.commit();
	}

	private ExtObjectContainer containerFor(final DrsProviderFixture fixture) {
		return ((Db4oReplicationProvider)fixture.provider()).getObjectContainer();
	}
	
	private long versionFor(final Data c1) {
		return objectInfoFor(c1).getCommitTimestamp();
	}

	private ObjectInfo objectInfoFor(final Data c1) {
		return containerFor(a()).ext().getObjectInfo(c1);
	}
	
	private void ods(final Object o) {
//		System.out.println(o);
	}

	public static void main(String[] args) {
		
		new ConsoleTestRunner(
				new DrsTestSuiteBuilder(
						new Db4oDrsFixture("db4o-a"),
						new Db4oDrsFixture("db4o-b"),
						PartialCollectionReplicationTestCase.class)).run();
	}
}

File: db4o.net/drs/Db4objects.Drs.Tests/Db4objects.Drs.Tests/Db4o/PartialCollectionReplicationTestCase.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

using System;
using System.Collections;
using System.Collections.Generic;
using Db4oUnit;
using Db4objects.Db4o.Events;
using Db4objects.Db4o.Ext;
using Db4objects.Db4o.Foundation;
using Db4objects.Drs.Db4o;
using Db4objects.Drs.Tests;
using Db4objects.Drs.Tests.Db4o;

namespace Db4objects.Drs.Tests.Db4o
{
	public class PartialCollectionReplicationTestCase : DrsTestCase
	{
		public class Data
		{
			private IList _children;

			private string _id;

			public Data(string id)
			{
				_id = id;
				_children = new ArrayList();
			}

			public virtual object Id()
			{
				return _id;
			}

			public virtual void Id(string id)
			{
				_id = id;
			}

			public virtual void Add(PartialCollectionReplicationTestCase.Data data)
			{
				_children.Add(data);
			}

			public override string ToString()
			{
				return "Data(" + _id + ", " + _children + ")";
			}
		}

		public virtual void TestNoReplicationForUntouchedElements()
		{
			PartialCollectionReplicationTestCase.Data root = new PartialCollectionReplicationTestCase.Data
				("root");
			PartialCollectionReplicationTestCase.Data c1 = new PartialCollectionReplicationTestCase.Data
				("c1");
			PartialCollectionReplicationTestCase.Data c2 = new PartialCollectionReplicationTestCase.Data
				("c2");
			root.Add(c1);
			root.Add(c2);
			Store(root, 1);
			IList<PartialCollectionReplicationTestCase.Data> created = ReplicateAllCapturingCreatedObjects
				();
			AssertData(created, "root", "c1", "c2");
			PartialCollectionReplicationTestCase.Data c3 = new PartialCollectionReplicationTestCase.Data
				("c3");
			root.Add(c3);
			Store(root, 2);
			c2.Id("c2*");
			c2.Add(new PartialCollectionReplicationTestCase.Data("c4"));
			IList<PartialCollectionReplicationTestCase.Data> updated = ReplicateAllCapturingUpdatedObjects
				();
		}

		// The following fails after cleaning references has been removed from #replicate(obj) 
		// assertData(updated, "c3", "root");
		private void AssertData(IEnumerable<PartialCollectionReplicationTestCase.Data> data
			, params string[] expectedIds)
		{
			Iterator4Assert.SameContent(expectedIds, Ids(data));
		}

		private IEnumerator Ids(IEnumerable<PartialCollectionReplicationTestCase.Data> data
			)
		{
			Collection4 ids = new Collection4();
			foreach (PartialCollectionReplicationTestCase.Data d in data)
			{
				ids.Add(d.Id());
			}
			return ids.GetEnumerator();
		}

		private IList<PartialCollectionReplicationTestCase.Data> ReplicateAllCapturingUpdatedObjects
			()
		{
			IList<PartialCollectionReplicationTestCase.Data> changed = new List<PartialCollectionReplicationTestCase.Data
				>();
			ListenToUpdated(changed);
			ListenToCreated(changed);
			ReplicateAll();
			return changed;
		}

		private IList<PartialCollectionReplicationTestCase.Data> ReplicateAllCapturingCreatedObjects
			()
		{
			IList<PartialCollectionReplicationTestCase.Data> created = new List<PartialCollectionReplicationTestCase.Data
				>();
			ListenToCreated(created);
			ReplicateAll();
			return created;
		}

		private void ListenToUpdated(IList<PartialCollectionReplicationTestCase.Data> updated
			)
		{
			EventRegistryFor(B()).Updated += new System.EventHandler<Db4objects.Db4o.Events.ObjectInfoEventArgs>
				(new _IEventListener4_97(this, updated).OnEvent);
		}

		private sealed class _IEventListener4_97
		{
			public _IEventListener4_97(PartialCollectionReplicationTestCase _enclosing, IList
				<PartialCollectionReplicationTestCase.Data> updated)
			{
				this._enclosing = _enclosing;
				this.updated = updated;
			}

			public void OnEvent(object sender, Db4objects.Db4o.Events.ObjectInfoEventArgs args
				)
			{
				object o = ((ObjectEventArgs)args).Object;
				if (o is PartialCollectionReplicationTestCase.Data)
				{
					updated.Add((PartialCollectionReplicationTestCase.Data)o);
				}
				this._enclosing.Ods(o);
			}

			private readonly PartialCollectionReplicationTestCase _enclosing;

			private readonly IList<PartialCollectionReplicationTestCase.Data> updated;
		}

		private void ReplicateAll()
		{
			Ods("BEGIN REPLICATION");
			ReplicateAll(A().Provider(), B().Provider());
			Ods("END REPLICATION");
		}

		private void ListenToCreated(IList<PartialCollectionReplicationTestCase.Data> created
			)
		{
			EventRegistryFor(B()).Created += new System.EventHandler<Db4objects.Db4o.Events.ObjectInfoEventArgs>
				(new _IEventListener4_115(this, created).OnEvent);
		}

		private sealed class _IEventListener4_115
		{
			public _IEventListener4_115(PartialCollectionReplicationTestCase _enclosing, IList
				<PartialCollectionReplicationTestCase.Data> created)
			{
				this._enclosing = _enclosing;
				this.created = created;
			}

			public void OnEvent(object sender, Db4objects.Db4o.Events.ObjectInfoEventArgs args
				)
			{
				object o = ((ObjectEventArgs)args).Object;
				if (o is PartialCollectionReplicationTestCase.Data)
				{
					created.Add((PartialCollectionReplicationTestCase.Data)o);
				}
				this._enclosing.Ods(o);
			}

			private readonly PartialCollectionReplicationTestCase _enclosing;

			private readonly IList<PartialCollectionReplicationTestCase.Data> created;
		}

		private IEventRegistry EventRegistryFor(IDrsProviderFixture fixture)
		{
			return EventRegistryFactory.ForObjectContainer(ContainerFor(fixture));
		}

		public virtual void TestCollectionUpdateDoesNotTouchExistingElements()
		{
			PartialCollectionReplicationTestCase.Data root = new PartialCollectionReplicationTestCase.Data
				("root");
			PartialCollectionReplicationTestCase.Data c1 = new PartialCollectionReplicationTestCase.Data
				("c1");
			PartialCollectionReplicationTestCase.Data c2 = new PartialCollectionReplicationTestCase.Data
				("c2");
			root.Add(c1);
			root.Add(c2);
			Store(root, 1);
			long c1Version = VersionFor(c1);
			long c2Version = VersionFor(c2);
			PartialCollectionReplicationTestCase.Data c3 = new PartialCollectionReplicationTestCase.Data
				("c3");
			root.Add(c3);
			Store(root, 2);
			Assert.IsGreater(0, VersionFor(c3));
			Assert.AreEqual(c1Version, VersionFor(c1));
			Assert.AreEqual(c2Version, VersionFor(c2));
		}

		private void Store(PartialCollectionReplicationTestCase.Data root, int depth)
		{
			IExtObjectContainer container = ContainerFor(A());
			container.Ext().Store(root, depth);
			container.Commit();
		}

		private IExtObjectContainer ContainerFor(IDrsProviderFixture fixture)
		{
			return ((IDb4oReplicationProvider)fixture.Provider()).GetObjectContainer();
		}

		private long VersionFor(PartialCollectionReplicationTestCase.Data c1)
		{
			return ObjectInfoFor(c1).GetCommitTimestamp();
		}

		private IObjectInfo ObjectInfoFor(PartialCollectionReplicationTestCase.Data c1)
		{
			return ContainerFor(A()).Ext().GetObjectInfo(c1);
		}

		private void Ods(object o)
		{
		}

		//		System.out.println(o);
		public static void Main(string[] args)
		{
			new ConsoleTestRunner(new DrsTestSuiteBuilder(new Db4oDrsFixture("db4o-a"), new Db4oDrsFixture
				("db4o-b"), typeof(PartialCollectionReplicationTestCase))).Run();
		}
	}
}
